rs/nns/integration_tests/src/cycles_minting_canister.rs

```diff
@@ -778,6 +784,255 @@ fn test_cmc_cycles_create_with_settings() {
     );
 }
 
+#[test]
+fn test_cmc_automatically_refunds_when_memo_is_garbage() {
+    // Step 1: Prepare the world.
+
+    // USER1 has some ICP (in their default subaccount).
+    let account_identifier = AccountIdentifier::new(*TEST_USER1_PRINCIPAL, None);
+    let initial_balance = Tokens::new(100, 0).unwrap();
+    let state_machine = state_machine_builder_for_nns_tests().build();
+    let nns_init_payloads = NnsInitPayloadsBuilder::new()
+        .with_ledger_account(account_identifier, initial_balance)
+        .build();
+    setup_nns_canisters(&state_machine, nns_init_payloads);
+
+    let assert_canister_statuses_fixed = |narrative_phase| {
+        assert_eq!(
+            btreemap! {
+                btreemap! { "status".to_string() => "running".to_string() } => 11,
+                btreemap! { "status".to_string() => "stopped".to_string() } => 0,
+                btreemap! { "status".to_string() => "stopping".to_string() } => 0,
+            },
+            fetch_int_gauge_vec(
+                state_machine.metrics_registry(),
+                "replicated_state_registered_canisters"
+            ),
+            "{}",
+            narrative_phase
+        );
+    };
+    // This will be called again later to verify that no canisters were added.
+    // Here, we just make sure that assert_canister_statuses_fixed has a correct
+    // understanding of how many canisters there are originally.
+    assert_canister_statuses_fixed("start");
+
+    let assert_balance = |nominal_amount_tokens: u64, fee_count: u64, narrative_phase: &str| {
```

Comment:
nit: could "narrative_phase" be renamed to "test_phase" or something?  There are a couple steps between "test" and "narrative" as concepts.